### DISCUSSION TITLE
map getter in assert stmt returns 'unknown value' instead of entry value

### DISCUSSION THREAD

**Discussion by jdonnerstag** [vlang/v#12694] [Ideas]
**V version:** V 0.2.4 52df19e
**OS:**  linux, Ubuntu 18.04.6 LTS (WSL)

**What did you do?**
```
fn test_map() ? {
	mut m := map[string]string{}
	m["a"] = r"c:\temp"
	assert m["a"] == r"c:\temp"
	assert m["a"] == os.join_path(r"c:\", "temp")
}
``` 

**What did you expect to see?**
All assert statements passing
 
**What did you see instead?**
```
> assert m['a'] == rosie.parser.os.join_path('c:\\', 'temp')
    Left value: *unknown value*
    Right value: c:/temp
```

The left value very clearly is a string, and `a` is a valid map entry. But the value shown and applied by assert is `*unknown value*`, and the test fails.

The workaround is to create a temp variable and compare this with the expected result. But it should not be necessary.

I have to say this is not 100% reproducable for me on different platforms and exact situations.

**Comment by spytheman** [comment]
The reason the assert is failing, is because os.join_path uses the current platform's path separator so on Ubuntu, that will be `/`, while on Windows, that would be `\`.

The `unknown value` is not an actual value, it is just a way for the assert statement to show you that it does not know how to unpack/display the expression (it only knows about certain limited expressions, and retrieving a value from a map is not yet one) -> that is a feature request, not a bug.

**Comment by jdonnerstag** [comment]
Are you sure? 

`m["a"] = r"c:\temp"`  assigns a string to the `a` map-entry. And in the assert statement, that is the left value. In the assert error message it is the left value showing `*unknown value*`. There is nothing to unpack, just a string.

The right value is made up of os.join_path() and that value is properly joined with the linux separator, and in fact properly shown in the error message. 

From a V/assert point of view, it is invoking a function (os.join_path) which returns a string. Nothing really to unpack.

A string as left value, and a string as right value.

**Comment by spytheman** [comment]
> Are you sure?

Yes, I am.

**Comment by spytheman** [comment]
The assert is triggered because of the mismatch, the shown value `*unknown value*` is because the assert does not know how to pretty print it. Source: I am the one who implemented the assert functionality.

**Comment by jdonnerstag** [comment]
> (it only knows about certain limited expressions, and retrieving a value from a map is not yet one) 

May be an extendable solution could be to leverage str() as in `"${mymap[2]}"`. This way it would even work for types unknown to the V compiler.
